crates/iceberg/src/arrow.rs

```diff
@@ -113,6 +143,405 @@ impl ArrowReader {
+        Ok(())
+    }
+
+    fn visit_always_true(&mut self) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn visit_always_false(&mut self) -> Result<Self::T> {
+        Ok(())
+    }
+
+    fn visit_unary(&mut self, predicate: &UnaryExpression<BoundReference>) -> Result<Self::T> {
+        self.bound_reference(predicate.term())?;
+        Ok(())
+    }
+
+    fn visit_binary(&mut self, predicate: &BinaryExpression<BoundReference>) -> Result<Self::T> {
+        self.bound_reference(predicate.term())?;
+        Ok(())
+    }
+
+    fn visit_set(&mut self, predicate: &SetExpression<BoundReference>) -> Result<Self::T> {
+        self.bound_reference(predicate.term())?;
+        Ok(())
+    }
+
+    fn bound_reference(&mut self, reference: &BoundReference) -> Result<Self::T> {
+        self.field_ids.push(reference.field().id);
+        Ok(())
+    }
+}
+
+struct PredicateConverter<'a> {
+    pub columns: &'a Vec<usize>,
+    pub projection_mask: ProjectionMask,
+    pub parquet_schema: &'a SchemaDescriptor,
+    pub column_map: &'a HashMap<i32, usize>,
+}
+
+fn get_arrow_datum(datum: &Datum) -> Box<dyn ArrowDatum> {
+    match datum.literal() {
+        PrimitiveLiteral::Boolean(value) => Box::new(BooleanArray::new_scalar(*value)),
+        PrimitiveLiteral::Int(value) => Box::new(Int32Array::new_scalar(*value)),
+        PrimitiveLiteral::Long(value) => Box::new(Int64Array::new_scalar(*value)),
+        PrimitiveLiteral::Float(value) => Box::new(Float32Array::new_scalar(value.as_f32())),
+        PrimitiveLiteral::Double(value) => Box::new(Float64Array::new_scalar(value.as_f64())),
+        _ => todo!("Unsupported literal type"),
+    }
+}
+
+impl<'a> BoundPredicateVisitor for PredicateConverter<'a> {
+    type T = Box<dyn ArrowPredicate>;
+    type U = usize;
+
+    fn visit_always_true(&mut self) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            |batch| Ok(BooleanArray::from(vec![true; batch.num_rows()])),
+        )))
+    }
+
+    fn visit_always_false(&mut self) -> Result<Self::T> {
+        Ok(Box::new(ArrowPredicateFn::new(
+            self.projection_mask.clone(),
+            |batch| Ok(BooleanArray::from(vec![false; batch.num_rows()])),
+        )))
+    }
+
+    fn visit_unary(&mut self, predicate: &UnaryExpression<BoundReference>) -> Result<Self::T> {
+        let term_index = self.bound_reference(predicate.term())?;
+
+        match predicate.op() {
+            PredicateOperator::IsNull => Ok(Box::new(ArrowPredicateFn::new(
+                self.projection_mask.clone(),
+                move |batch| {
+                    let column = batch.column(term_index);
+                    is_null(column)
+                },
+            ))),
+            PredicateOperator::NotNull => Ok(Box::new(ArrowPredicateFn::new(
+                self.projection_mask.clone(),
+                move |batch| {
+                    let column = batch.column(term_index);
```

Comment:
This maybe incorrect for nested column, I think maybe we should either return projection_mask for each leave column, or implement a general purpose flatten method for struct array.

Reply:
The Parquet reader API will flatten required columns based on the projection_mask we provide. I.e., If the projection mask selects one nested column `a.b`, it will be the first column of the record batch when calling `evaluate` of `ArrowPredicate` API.

Reply:
I did some test but it seems that it doesn't work like this, say the schem is like following:
```
message {
  struct a {
    int b
  }
}
```
And we pass `ProjectionMask::leaves([0])`, it will return struct array, so `batch.column[term_index]` will return `StructArray`

Reply:
Hmm, that's what I got from reading the Parquet code. Let me take further look and test.

Reply:
Just wrote a test to test it in arrow-rs. Yea, for a nested column such as `struct.a`, the batch passed to `evaluate` contains a struct array with the column `a`. Looks like Parquet will project the requested column indices and construct the upper nested type (i.e., `struct`) before passing to `evaluate`,

Reply:
Added the test to arrow-rs to clarify its usage: https://github.com/apache/arrow-rs/pull/5600

Reply:
I think to have a flatten method for struct array sounds more simple way. I'm looking in arrow-rs to see if there is existing one, if not, we need to implement it here.

Reply:
> This maybe incorrect for nested column, I think maybe we should either return projection_mask for each leave column, or implement a general purpose flatten method for struct array.

I tried to change to return projection_mask for each leave column, it is pretty straightforward to implement. Please let me know if it looks good to you. Thanks.